tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11299,6 +11299,62 @@ bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
   return false;
 }
 
+void ValidateWaveSize(clang::Sema *S, FunctionDecl *FD) {
+  const auto *SM =
+      hlsl::ShaderModel::GetByName(S->getLangOpts().HLSLProfile.c_str());
+
+  const llvm::StringRef &entryName = S->getLangOpts().HLSLEntryFunction;
+  StringRef functionName = "";
+  if (FD->getIdentifier())
+    functionName = FD->getName();
+  bool isEntry = false;
+  bool isNode = false;
+  bool isCS = false;
+
+  isEntry =
+      !SM->IsLib() &&
+      FD->getDeclContext()->getDeclKind() == Decl::Kind::TranslationUnit &&
+      functionName == entryName;
+
+  auto *shaderAttribute = FD->getAttr<HLSLShaderAttr>();
+  StringRef Literal = shaderAttribute->getStage();
+  DXIL::ShaderKind Stage = ShaderModel::KindFromFullName(Literal);
```

Comment:
I've made the parent function pass down `ShaderModel` and the `HLSLShaderAttr`. I don't think it made sense to pass the `ShaderKind` because it is possible it won't exist, and it's difficult to express a non-existent shaderkind without using invalid, which isn't an accurate representation.

For now, I haven't passed down whether or not this is an active entry point, some further discussion will need to be had.